### PR TITLE
Feature/buffer latest data

### PR DESCRIPTION
**Pull request by yunstanford**
This PR includes:

1. modify _Memcache class in carbon
2. push back new metrics dropped by carbon-cache when reaching new_metrics_create limit.
3. apply changes to unit tests.
4. add protocol to return storage-schema.
5. add protocol to do carbon-query-precheck

etc.

### CODE REVIEW

File: lib/carbon/cache.py [hunk 1]
```diff
@@ -116,12 +119,28 @@ def counts(self):
     return [(metric, len(datapoints)) for (metric, datapoints) in self.items()]
 
   @property
+  def unflush_counts(self):
+    return [(metric, self._count_unflushed(datapoints)) for (metric, datapoints) in self.items()]
```
Comment (toumorokoshi) [inline lib/carbon/cache.py]: can this not be an integer we count? seems unperformant to pay the cost of counting every property call.

File: lib/carbon/cache.py [hunk 2]
```diff
@@ -137,11 +156,42 @@ def drain_metric(self):
...
+
+    return (metric, datapoints)
 
   def get_datapoints(self, metric):
     """Return a list of currently cached datapoints sorted by timestamp"""
-    return sorted(self.get(metric, {}).items(), key=by_timestamp)
+    # let's keep old interface, strip out is_flushed before return
+
+    datapoints = sorted(self.get(metric, {}).items(), key=by_timestamp)
+    return [(timestamp, value) for (timestamp, (value, is_flushed)) in datapoints]
```
Comment (toumorokoshi) [inline lib/carbon/cache.py]: probably unnecessary, but you could call sorted on get_unsorted_datapoints here. some DRY.

File: lib/carbon/protocols.py [hunk 3]
```diff
@@ -207,7 +210,7 @@ def stringReceived(self, rawRequest):
       datapointsByMetric = {}
       metrics = request['metrics']
       for metric in metrics:
-        datapointsByMetric[metric] = MetricCache.get(metric, {}).items()
+        datapointsByMetric[metric] = MetricCache.get_unsorted_datapoints(metric)
```
Comment (toumorokoshi) [inline lib/carbon/protocols.py]: nice encapsulation.

**Comment by toumorokoshi** [comment]
LGTM

File: lib/carbon/cache.py (hunk 1 above)
Comment (yunstanford) [inline lib/carbon/cache.py]: Maintaining a large map {"metric": unflushed_count } ? i can do that as well.
 As we are introducing the "is_flushed" field, we are going to pay the cost anyway.

**Comment by toumorokoshi** [comment]
LGTM

**Pull request merged**